libparsers/src/xml/stations_and_request_stops.cpp

```diff
@@ -0,0 +1,430 @@
+					util::lower(behaviour);
+				}
+				catch (std::exception const& e) {
+					add_error(errors, filename, 0, e.what());
+					behaviour = "normalbrake"s;
+				}
+				auto const parsed_behaviour = text_mapping.find(behaviour);
+
+				if (parsed_behaviour != text_mapping.end())
+					return parsed_behaviour->second;
+
+				add_error(errors, filename, 0, "Error: <AIBehaviour> given invalid behaviour");
+				return request_stop_marker::behaviour::normalbrake;
+			}
+
+			sub_message parse_sub_messages_nodes(rapidxml_ns::xml_node<char>* parent_node) {
+				auto* early_node = parent_node->first_node("early", 0, false);
+				auto* on_time_node = parent_node->first_node("ontime", 0, false);
+				auto* late_node = parent_node->first_node("late", 0, false);
+				sub_message sub_nodes;
+
+				// There are no sub nodes so return a struct with values of "parent_node".
+				if (early_node == nullptr && on_time_node == nullptr && late_node == nullptr) {
+					sub_nodes.early = parent_node->value();
+					sub_nodes.ontime = parent_node->value();
+					sub_nodes.late = parent_node->value();
+					return sub_nodes;
+				}
+
+				if (early_node != nullptr) {
+					sub_nodes.early = early_node->value();
+				}
+				else {
+					sub_nodes.early = std::string{};
+				}
+
+				if (on_time_node != nullptr) {
+					sub_nodes.ontime = on_time_node->value();
+				}
+				else {
+					sub_nodes.ontime = std::string{};
+				}
+
+				if (late_node != nullptr) {
+					sub_nodes.late = late_node->value();
+				}
+				else {
+					sub_nodes.late = std::string{};
+				}
+
+				return sub_nodes;
+			}
+
+			probabilities parse_probabilities_nodes(std::string const& filename,
+			                                        rapidxml_ns::xml_node<char>* parent_node,
+			                                        errors::multi_error_t& errors) {
+				auto* early_node = parent_node->first_node("early", 0, false);
+				auto* on_time_node = parent_node->first_node("ontime", 0, false);
+				auto* late_node = parent_node->first_node("late", 0, false);
+				probabilities sub_nodes;
+
+				// There are no sub nodes so return a struct with values of "parent_node".
+				if (early_node == nullptr && on_time_node == nullptr && late_node == nullptr) {
+					std::intmax_t no_sub_values;
+					try {
+						no_sub_values = util::parse_loose_integer(parent_node->value());
+					}
+					catch (std::exception const& e) {
+						no_sub_values = 0;
+						add_error(errors, filename, 0, e.what());
+					}
+					if (no_sub_values < 0 || no_sub_values > 100) {
+						add_error(errors, filename, 0,
+						          "<Probability> has to be between 0 and 100 but found"
+						              + no_sub_values);
+						no_sub_values = 0;
+					}
+					sub_nodes.early = gsl::narrow<std::uint8_t>(no_sub_values);
+					sub_nodes.ontime = gsl::narrow<std::uint8_t>(no_sub_values);
+					sub_nodes.late = gsl::narrow<std::uint8_t>(no_sub_values);
+					return sub_nodes;
+				}
+
+				if (early_node != nullptr) {
```

Comment:
These three ifs can be refactored into a lambda taking a reference. When you get to dealing with this, tell me and we'll discuss.